File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.py
class MyQueue(object):
    # stack: LIFO
    # queue: FIFO
    def __init__(self):
        self.instack = []
        self.outstack = []

    def push(self, x):
        self.instack.append(x)

    def pop(self):
        self.peek()
        return self.outstack.pop()

    def peek(self):
        if not self.outstack:
            while self.instack:
                self.outstack.append(self.instack.pop())
        return self.outstack[-1]

    def empty(self):
        return not self.instack and not self.outstack
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.cpp
class MyQueue {
public:
    MyQueue() {    }
    stack<int> instack;
    stack<int> outstack;
    
    void push(int x) {
        instack.push(x);
    }
    
    int pop() {
        this->peek();
        int x = outstack.top();
        outstack.pop();
        return x;
    }
    
    int peek() {
        if (outstack.empty()) {
            while (!instack.empty()) {
                int x = instack.top();
                instack.pop();
                outstack.push(x);
            }
        }

        return outstack.top();
    }
    
    bool empty() {
        if (instack.empty() && outstack.empty()) return true;
        return false;
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */